pkg/apis/seedmanagement/v1alpha1/types_managedseed.go

```diff
@@ -0,0 +1,216 @@
+	// Standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the ManagedSeed.
+	Spec ManagedSeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Most recently observed status of the ManagedSeed.
+	Status ManagedSeedStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ManagedSeedList is a list of ManagedSeed objects.
+type ManagedSeedList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list object metadata.
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Items is the list of ManagedSeeds.
+	Items []ManagedSeed `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ManagedSeedSpec is the specification of a ManagedSeed.
+type ManagedSeedSpec struct {
+	// Shoot is the Shoot that will be registered as Seed.
+	// +optional
+	Shoot *Shoot `json:"shoot,omitempty" protobuf:"bytes,1,opt,name=shoot"`
+	// Seed describes the Seed that will be registered.
+	// Either Seed or Gardenlet must be specified.
+	// +optional
+	Seed *SeedTemplateSpec `json:"seed,omitempty" protobuf:"bytes,2,opt,name=seed"`
+	// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+	// +optional
+	Gardenlet *Gardenlet `json:"gardenlet,omitempty" protobuf:"bytes,3,opt,name=gardenlet"`
+}
+
+// Shoot identifies the Shoot that will be registered as Seed.
+type Shoot struct {
+	// Name is the name of the Shoot that will be registered as Seed.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// APIServer specifies certain kube-apiserver parameters of the Shoot that will be registered as Seed.
+	// +optional
+	APIServer *APIServer `json:"apiServer,omitempty" protobuf:"bytes,2,opt,name=apiServer"`
+}
+
+// SeedTemplateSpec describes the data a Seed should have when created from a template.
+type SeedTemplateSpec struct {
+	// Standard object metadata.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the desired behavior of the Seed.
+	// +optional
+	Spec gardencorev1beta1.SeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+type Gardenlet struct {
+	// Deployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+	// the image, which bootstrap mechanism to use (bootstrap token / service account), etc.
+	// +optional
+	Deployment *GardenletDeployment `json:"deployment,omitempty" protobuf:"bytes,1,opt,name=deployment"`
+	// Config is the GardenletConfiguration used to configure gardenlet.
+	// +optional
+	Config *gardenletconfigv1alpha1.GardenletConfiguration `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`
```

Comment:
Hm, isn't this a classical issue with API versioning that is supposed to be solved with versioned APIs? I mean, when we need to integrate a new version of `GardenletConfiguration` we could add a new version of `ManagedSeed`. We also have a lot of references to e.g. `gardenercorev1beta1` here and elsewhere, not the mention the Kubernetes APIs, how are we supposed to handle version changes in these? 
Using `RawExtension` has its drawbacks, e.g. validation becomes more complex. Also, how do you decide what version of the `GardenletConfigration` is ok to be accepted?

Reply:
> I mean, when we need to integrate a new version of `GardenletConfiguration` we could add a new version of `ManagedSeed`

That's actually, what I would try to avoid. If we use `RawExtension`, we will be able to easily support the new version of the config API without adding a new seedmanagement API version.

> We also have a lot of references to e.g. `gardenercorev1beta1` here and elsewhere, not the mention the Kubernetes APIs, how are we supposed to handle version changes in these?

The difference between the gardenlet config and the gardener core/k8s types is, that the gardenlet config is actually an external API and not directly coupled to the seedmanagement API.

> not the mention the Kubernetes APIs, how are we supposed to handle version changes in these?

Well, the k8s APIs are mostly in GA (`v1`), so there is a good guarantee of stability/compatibility. If they will ever introduce `v2`, it will be a breaking change, so one would also definitely need a new version of the gardener APIs if they reference some of the changed types. The difference to the gardenlet config API is, that it is only in `v1alpha1` and therefore has basically no stability/compatibility guarantee. So it is not so unlikely to change in the future. If we couple the seedmanagement to the gardenlet config API, we will need to keep them in lockstep, which is probably not what we want to do.

> Using `RawExtension` has its drawbacks, e.g. validation becomes more complex. Also, how do you decide what version of the `GardenletConfigration` is ok to be accepted?

It isn't to complex, I guess. The seedmanagement validation can simply decode the config section, detect the version and call the corresponding validation. At first, only `v1alpha1` would be supported. But as soon as we have a new version of the config API, we can easily support it, without changing the seedmanagement API.

Reply:
OK, thanks, I understand better what you mean now. I still have some concerns. How about merging with parent? If the parent GardenletConfiguration is e.g. `v1alpha1` and the one in the resource is e.g. `v1beta1`, how should they be merged? I guess we should use the "unversioned" configuration in the controller code already today to perform the merging?

Reply:
That's a valid point. Yes, merging should probably be done with the internal version.

Reply:
@rfranzke @timebertt Something that came to mind: if we use `RawExtension`, having a new `seedmanagement` group is no longer mandatory. Originally it was introduced to break the circular dependency btw `core/v1beta1` and `config/v1alpha1`. With this circular dependency gone, we could move `ManagedSeed` and all related resources to `core` and get rid of `seedmanagement` if we wanted to. WDYT?

Reply:
I am fine with both and I don't have a clear preference, TBH. Probably one API group only would be easier as we haven't grouped our other resources thoroughly in the past. OTOH, as all of this is new (`ManagedSeed`, etc.) and as there is more to come (`ManagedSeedSet`, autoscaler resources, etc.), it is now a good opportunity for a new API group if a we want to follow this path.
Do you guys have preferences?

Reply:
I actually like having a new API group for this, especially with the other resources that will be introduced later on. So I have a slight preference to keep it :)

Reply:
OK, I will switch to `RawExtension` but keep the new API group.